API for AutogradContext

malvika2147 opened the pull request:
### NOTE: All the PRs in this stack have been moved to #23572, abandoning this PR

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23184 API for AutogradContext**
* #23020 Support custom autograd functions in C++

Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
API for AutogradContext

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -41,11 +56,37 @@ struct CFunction {
   std::unordered_set<at::TensorImpl*> dirty_inputs;
+  std::vector<torch::autograd::SavedVariable> saved_variables;
+
+  std::shared_ptr<Function> grad_fn;
```
Is it ok in cpp to have a shared_ptr that contains an object that contains this shared pointer?
The Function owns the context here and then context seems to own the Function as well. Or I got lost somewhere?

ezyang replied:
Yeah, this is probably causing a refcycle.

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -41,11 +56,37 @@ struct CFunction {
 struct AutogradContext {
-  std::vector<torch::autograd::SavedVariable> saved_variables;
+  // Can be used to save non-variable data for backward()
+  std::unordered_map<std::string, at::IValue> save;
```
Not sure about this name. Maybe "user_data"? I'm really not sure if it's better.

albanD commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -18,19 +19,33 @@ TORCH_API variable_list _wrap_outputs(
 // class MyFunction : public CFunction<MyFunction> {
 //   public:
-//   static variable_list forward(AutogradContext *ctx, int n, Variable var);
+//   static variable_list forward(AutogradContext *ctx, int n, Variable var) {
```
It would be very nice if the example also showed where and when to use `mark_dirty`  (and `mark_non_differentiable` but this one is more obvious).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -18,33 +20,80 @@ TORCH_API variable_list _wrap_outputs(
+  const std::unordered_set<at::TensorImpl*>& get_dirty() const;
+  const std::unordered_set<at::TensorImpl*>& get_non_differentiable() const;
+
+private:
```
OK good, I like the private specifier better. Actually, I wonder if `saved_data` shouldn't be private too. We could probably give a more user friendly way of setting and getting.

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -18,33 +20,80 @@ TORCH_API variable_list _wrap_outputs(
 //
 // For example:
-// class MyFunction : public CFunction<MyFunction> {
+// class MyFunction : public Function<MyFunction> {
```
nit: Technically this should have been in your previous patch ;)

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -18,33 +20,80 @@ TORCH_API variable_list _wrap_outputs(
+  // Saves the list of variables for a future call to backward(). This
+  // should be called at most once from inside of forward().
+  void save_for_backward(const variable_list &to_save);
+  // Marks variables in the list as modified in an in-place opertaion. This
```
nit: operation

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -18,33 +20,80 @@ TORCH_API variable_list _wrap_outputs(
   std::unordered_set<at::TensorImpl*> dirty_inputs;
+  std::vector<torch::autograd::SavedVariable> saved_variables;
+
+  std::weak_ptr<Node> grad_fn;
```
The weak pointer here is a little bit of a code smell, so it would be helpful if there were a comment explaining why we need to have it here.

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -18,33 +20,80 @@ TORCH_API variable_list _wrap_outputs(
+  // Marks outputs in the list as not requiring gradients. This should be called
+  // at most once from inside of forward() and all arguments should be outputs.
+  void mark_non_differentiable(const variable_list &outputs);
+  void clear_saved();
```
Does this really need to be public? Maybe we can make it private and friend CppNode.

ezyang commented:
FYI, if it's easier for you, I'm OK if you scrunch all of these into one PR. (You don't have to, I'm just offering it as an option.)

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -100,4 +100,50 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+variable_list AutogradContext::get_saved_variables() const {
+  variable_list saved;
+  saved.reserve(saved_variables.size());
+  auto ptr = grad_fn.lock();
```
You need to assert that `ptr` is non-null here, I think.

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -100,4 +100,50 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+  saved_variables.clear();
+}
+
+variable_list AutogradContext::get_saved_variables() const {
```
In the original implementation `unpack_saved_variables`, there is an assert that the function does not have freed buffers. You need to do that test here. Buffers being freed is one of the main situations when the `ptr` below could be null.

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -100,4 +100,50 @@ variable_list _wrap_outputs(const variable_list &input_vars,
+}
+
+void AutogradContext::clear_saved() {
+  saved_variables.clear();
```
So you would then set `has_freed_buffers = true` here

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -100,4 +100,50 @@ variable_list _wrap_outputs(const variable_list &input_vars,
 }
 
+void AutogradContext::save_for_backward(const variable_list &to_save) {
+  saved_variables.clear();
```
This design is a bit iffy. In the original version, `_save_variables` (btw, remember the comment I made about cross-referencing the implementations before and after? Please do that here too!), this function is not actually a public user visible method; it is called after application of the python function to read out the `to_save` member field (which the user actually wrote) and save them in a more durable form. But here, you've defined it as a user visible function, which ostensibly is called by implementations of AutogradContext. But now there is a problem: you haven't wrapped the output variable yet, so `(var.grad_fn().get() == ptr.get())` will always be false! That's no good.

ezyang reviewed (dismissed):
Some mild correctness issues

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

Commit:
Update on "API for AutogradContext"


Summary:
Both `forward()` and `backward() `accept a AutogradContext* which can be used to share data between them. 
Variables can be saved in the context using `save_for_backward()` and other data can be saved in the map `save` in the form of `<std::string, at::IValue>` pairs. Variables saved in forward can be accessed with `get_saved_variables()`.

Example usage:
```
class MyFunction : public CFunction<MyFunction> {
  public:
  static variable_list forward(AutogradContext *ctx, int n, Variable var) {
     // Save data for backward in context
     ctx->save["n"] = n;
     return std::vector<Variable>({var});
  }

  static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
     // Use data saved in forward
     auto n = ctx->save["n"].toInt();
     return std::vector<Variable>({grad_output[0]*n});
  }
};

```

Also AutogradContext has methods to mark outputs as non differentiable and mark inputs as dirty similar to the [Python API](https://github.com/pytorch/pytorch/blob/ff23a02ac4fdf1fe76d5b24666333f1ea0a918b7/torch/autograd/function.py#L26).

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -100,4 +100,56 @@ variable_list _wrap_outputs(const variable_list &input_vars,
 }
 
+void AutogradContext::save_for_backward(const variable_list &to_save) {
+  to_save_ = to_save;
```
nit: Since you are always going to do a copy, it is better to take `variable_list` by value. This means that if a user provides an rvalue reference, we can just move it into `to_save_`.

ezyang replied:
(inside the body, you would then do a move into `to_save_`)
